# Benchmarking Harpia Module for Annotat3d Image Processing and Segmentation

This notebook benchmarks the **Harpia module**, the backend engine for Annotat3d, a tool designed for advanced image processing and segmentation tasks. The benchmarking process evaluates two key performance metrics:

1. **Accuracy**: Compares the results of Harpia’s image processing and segmentation functions with standard open-source tools like `scikit-image`, ensuring reliability and accuracy that meet industry standards.

2. **Execution Time**: Assesses runtime efficiency by measuring the execution times of Harpia’s functions relative to baseline tools, especially when handling large or complex images.

The benchmark is conducted on both **2D and 3D images** of various sizes, including very large images, to reflect Sirius demands in high-resolution and volumetric data efficient processing. This notebook provides a step-by-step analysis, including data setup, function execution, and results comparison, to evaluate Harpia's capability to handle a range of image dimensions and formats.


In [ ]:
import os
import pandas as pd
import glob

def read_csv_files():
    # Get the current working directory
    current_dir = os.getcwd()
    print("Searching in:", current_dir)

    # Recursively find all CSV files in subdirectories
    csv_files = glob.glob(os.path.join(current_dir, "**", "*.csv"), recursive=True)

    if not csv_files:
        print("No CSV files found.")
        return None

    # Read all CSV files into a list of DataFrames
    dfs = [pd.read_csv(file) for file in csv_files]

    # Optionally, concatenate all DataFrames into one
    df_all = pd.concat(dfs, ignore_index=True)

    print(f"Loaded {len(csv_files)} CSV files.")
    return df_all

# Example usage
df = read_csv_files()


Searching in: /ibira/lnls/labs/tepui/home/camila.araujo/work/harpia/tests_python/benchmark
Loaded 110 CSV files.


In [ ]:
df['Image Dimensions'].unique()

array(['(1964, 2048, 2048)', '(1520, 1520, 600)', '(100, 207, 190)',
       '(2048, 2052, 2052)'], dtype=object)

In [ ]:
df[df['Image Dimensions'] == '(2048, 2052, 2052)'] #big image

,Operation,Machine,Gpus,gpuMemory,Module Time (s),Scikit-Image Time (s),Time Ratio,Repetitions,Mean Squared Error,Pixel Accuracy (%),Image Data Type,Image Size (MB),Image Dimensions
78,Closing 3D binary,harriet,1,0.2,28.519186,339.915684,11.9,10,0.0,100.0,uint32,32896.1,"(2048, 2052, 2052)"
79,Dilation 3D binary,harriet,1,0.2,6.069428,142.815209,23.5,10,0.0,100.0,int16,16448.1,"(2048, 2052, 2052)"
80,Dilation 3D binary,harriet,1,0.2,16.472503,167.151651,10.1,10,0.0,100.0,uint32,32896.1,"(2048, 2052, 2052)"
81,Closing 3D grayscale,harriet,1,0.2,25.848475,551.987434,21.4,10,0.0,100.0,float32,32896.1,"(2048, 2052, 2052)"
82,Closing 3D grayscale,harriet,1,0.2,26.245183,571.965393,21.8,10,0.0,100.0,uint32,32896.1,"(2048, 2052, 2052)"
83,Smoothing 3D binary,harriet,1,0.2,18.542238,646.361681,34.9,10,0.0,100.0,int32,32896.1,"(2048, 2052, 2052)"
84,Dilation 3D grayscale,harriet,1,0.2,25.374762,311.657135,12.3,10,0.0,100.0,uint32,32896.1,"(2048, 2052, 2052)"
85,Dilation 3D binary,harriet,1,0.2,6.095893,139.019403,22.8,10,0.0,100.0,uint16,16448.1,"(2048, 2052, 2052)"
86,Closing 3D binary,harriet,1,0.2,7.181724,293.469370,40.9,10,0.0,100.0,int16,16448.1,"(2048, 2052, 2052)"
87,Erosion 3D binary,harriet,1,0.2,16.463477,179.271287,10.9,10,0.0,100.0,uint32,32896.1,"(2048, 2052, 2052)"


In [ ]:
df[df['Image Dimensions'] == '(100, 207, 190)'] #small image

,Operation,Machine,Gpus,gpuMemory,Module Time (s),Scikit-Image Time (s),Time Ratio,Repetitions,Mean Squared Error,Pixel Accuracy (%),Image Data Type,Image Size (MB),Image Dimensions
36,Smoothing 3D binary,harriet,1,0.2,0.010380,0.226694,21.8,10,0.0,100.0,uint32,15.0,"(100, 207, 190)"
37,Opening 3D binary,harriet,1,0.2,0.004046,0.086353,21.3,10,0.0,100.0,uint8,3.8,"(100, 207, 190)"
38,Opening 3D binary,harriet,1,0.2,0.004510,0.088112,19.5,10,0.0,100.0,int8,3.8,"(100, 207, 190)"
39,Erosion 3D grayscale,harriet,1,0.2,0.005448,0.107222,19.7,10,0.0,100.0,uint32,15.0,"(100, 207, 190)"
40,Closing 3D binary,harriet,1,0.2,0.003869,0.088800,23.0,10,0.0,100.0,int8,3.8,"(100, 207, 190)"
41,Dilation 3D grayscale,harriet,1,0.2,0.006155,0.107939,17.5,10,0.0,100.0,uint32,15.0,"(100, 207, 190)"
42,Opening 3D binary,harriet,1,0.2,0.004679,0.087823,18.8,10,0.0,100.0,int16,7.5,"(100, 207, 190)"
43,Closing 3D grayscale,harriet,1,0.2,0.007507,0.190962,25.4,10,0.0,100.0,float32,15.0,"(100, 207, 190)"
44,Opening 3D binary,harriet,1,0.2,0.006745,0.094066,13.9,10,0.0,100.0,uint32,15.0,"(100, 207, 190)"
45,Erosion 3D binary,harriet,1,0.2,0.002882,0.045024,15.6,10,0.0,100.0,int8,3.8,"(100, 207, 190)"


## Erro na operação cudaMemcpy(deviceToHost)

O parâmetro gpuMemory controla o quanto da memorória livre da gpu pode ser usado para alocar a imagem de entrada e suas cópias. A parir desse parâmetro, da memória livre e do númeor de cópias do input usado pela função, calcula-se o tamanho máximo do chunk. 
Em testes locais, usando a gpu do notebook, é possível operar com gpuMemory próximo de 1. É precisa fechar todas as demais janelas e manter apenas o terminal aberto para que o sistema operacional não mate o processo (pois a gpu também é usada para renderizar a interface gráfica e o sistema fica sobrecarregado).
Nos testes remotos feitos na Harriet, não é possível aumentar o parâmetro gpuMemory. Para a imagem média (1520, 1520, 600) o valor máximo da gpuMemory que é possível executar o código é 0.41. Para a imagem grande (2048, 2052, 2052) esse valor cai para 0.2, podendo variar a depender da operação e do tipo de dado.
Especialmente para a imagem grande (2048, 2052, 2052) e operações de erosão e dilatação em dados uint8 surgiam erros de 'segmentation fault' para qualquer valor de gpuMemory. Nos demais tipos de dados, surgia erro recorrente na função cudaMemcpy(deviceToHost), especificamente quando copiava-se o resultado final de algum chunk do device para o host a partir de um determinado valor de gpuMemory.
Foram identificadas duas hipóteses de problema:
1) **Dados não emparelhados na memória** - como o tipo de dado uint8 ocupa apenas 1 byte, poderia estar alocado na memória fora do padrão esperado pelas funções cuda e, por isso, surgia o erro 'segmentation fault'.
**Proposta de solução:** alocar os dado no host para testes com funções cuda (cudaMallocHost), pois a função lidaria internamente com esse problema.

2) **Taxa de transferência de dados maior do que a permitida para a função cudaMemcpy(deviceToHost)** 
**Proposta de solução:** copiar os dados em chunks ainda menores de volta para o host.

Foram realizados testes implementando cada uma dessas soluções separadamente e juntas para imagens uint8 e uint16 e os valores de gpuMemory 0.4 e 0.9

### Resultados:

#### CudaMallocHost 
**uint8:**  Funcionou para gpuMemory 0.4 e 0.9.
**uint16:** Não funcionou para gpuMemory 0.4, 0.9 não foi testado.

#### CopyChunks
**uint8:** Não funcionou para gpuMemory 0.4 nem 0.9.
**uint16:** Funcionou para gpuMemory 0.4, e não funcionou para 0.9.

#### CudaMallocHost + CopyChunks 
**uint8:**  Não funcionou para gpuMemory 0.4 nem 0.9.
**uint16:** Funcionou para gpuMemory 0.4 e 0.9

### Conclusão
As soluçoes foram inconclusivas, porque funcionavam para um tipo de dado e para outro não. Não foram implementadas ainda.